src/Language/PureScript/TypeChecker/Unify.hs

```diff
@@ -176,8 +176,8 @@ unifyRows r1 r2 = sequence_ matches *> uncurry unifyTails rest where
     rest' <- freshTypeWithKind =<< elaborateKind (TUnknown a u1)
     solveType u1 (rowFromList (sd2, rest'))
     solveType u2 (rowFromList (sd1, rest'))
-  unifyTails _ _ =
-    throwError . errorMessage $ TypesDoNotUnify r1 r2
+  unifyTails r1' r2' =
+    throwError . errorMessage $ TypesDoNotUnify (rowFromList r1') (rowFromList r2')
```

Comment:
A different approach to consider here would be:
```suggestion
  unifyTails _ _ = do
    subst <- gets checkSubstitution
    throwError . errorMessage $ TypesDoNotUnify (substituteType subst r1) (substituteType subst r2)
```

This results in slightly different error messages—the upshot, thanks to row-diffing code that lives in the pretty-printer for types, is that in cases like 3701 the `...` appears when printing the rows that don't match with the `substituteType` approach but not with this approach.

Personally, I think having the `...` might be clearer. What do you think?

Reply:
I will try it when I have more free time than I have now. :)

Reply:
See https://github.com/purescript/purescript/pull/4421/commits/d4cf7215379239190d77f2be11b27fd040447fe4 for what this looks like. To summarize, while it adds the `...` in case 3701, it obfuscates the result for cases like `NestedRecordTypeError`. 

I initially thought we could add a hint before I realized that a hint is already added. The reason why it doesn't appear in this case is because the `ErrorUnifyingTypes` hint is stripped via `stripRedundantHints` when the error is `TypesDoNotUnify`. 

We could stop stripping that hint, but I don't think that tradeoff is worth it in this context.

Reply:
>  I would have thought that getting the current substitution would have filled in all those unknowns.

If I had to guess, I'd say it's because we haven't solved those _yet_. We run `parU matches id`, which iterates through the list in left-to-right order. If `a` is the first label, we haven't done `unifyTypes` on the other labels yet, so some unknowns may not have been solved yet. Thus, when the error for `a` is thrown, labels, `d`, `e`, and `f`, are still unknown.

Reply:
Nope, it's the double `parU`. If I unify them with
```
void $ parU (matches ++ [ uncurry unifyTails rest ]) id
```
then I get something closer to what I was expecting.

Downside of *that* (she swallowed the spider to catch the fly) is that, still in `NestedRecordTypeError`, now the `unifyTails` error message contains information redundant with the other error raised when aligning rows. I have a plan for that; I'll post when I'm happy with it.

Reply:
Jordan, take a look at the commit I just pushed? It's a little more baroque than simply using `parU`, but I think it improves the error messages.